server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java

```diff
@@ -297,10 +300,13 @@ private ClusterState applyRequest(ClusterState currentState, PutMappingClusterSt
                 final Index index = indexMetaData.getIndex();
                 final MapperService mapperService = indexMapperServices.get(index);
 
-                // If the user gave _doc as a special type value or if they are using the new typeless APIs,
-                // then we apply the mapping update to the existing type. This allows to move to typeless
-                // APIs with indices whose type name is different from `_doc`.
-                String typeForUpdate = mapperService.resolveDocumentType(mappingType); // the type to use to apply the mapping update
+                // If the _type name is _doc and there is no _doc top-level key then this means that we
+                // are handling a typeless call. In such a case, we override _doc with the actual type
+                // name in the mappings. This allows to use typeless APIs on typed indices.
+                String typeForUpdate = mappingType; // the type to use to apply the mapping update
+                if (isMappingSourceTyped(mapperService, mappingUpdateSource, request.type()) == false) {
```

Comment:
To check I understand: on an index with a custom type, issuing a typeless 'index' requests that updates the mappings will still work because `request.type()` will not be equal to the type key in `mappingUpdateSource`?

Reply:
The situation is different on 7.x and 6.x. On 7.x, a typeless index request on an index that has types triggers the creation of a put-mapping request for the actual type of the mapping, so we are good. This is different on 6.x which doesn't have logic to make typeless index requests work on typed indices, so doing it would actually create a typed put-mapping call with `_doc` as a type name. This is how I found about this issue when backporting the change that deals with merging mappings and templates.